tempesta_fw/http.c

```diff
@@ -1070,23 +1126,48 @@ tfw_http_adjust_resp(TfwHttpResp *resp)
 	return 0;
 }
 
+/*
+ * Depending on results of processing of a request, either send the request
+ * to an appropriate server, or return the cached response. If none of that
+ * can be done for any reason, return HTTP 404 error to the client.
+ */
 static void
 tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp, void *data)
 {
-	TfwConnection *conn = data;
+	int r;
 
 	if (resp) {
 		/*
 		 * We have prepared response, send it as is.
 		 * TODO should we adjust it somehow?
 		 */
-		tfw_connection_send(conn, (TfwMsg *)resp);
+		tfw_connection_send(req->conn, (TfwMsg *)resp);
 	} else {
-		if (tfw_http_adjust_req(req))
+		/* Dispatch request to an appropriate server. */
+		TfwConnection *conn = tfw_sched_get_srv_conn((TfwMsg *)req);
+		if (!conn)
+			goto send_404;
+		r = tfw_http_sticky_req_process((TfwHttpMsg *)req);
+		if (r < 0) {
+			goto send_404;		/* Should it be an HTTP 502? */
+		} else if (r > 0) {
+			/* Response sent, nothing to do */
 			return;
-		/* Send the request to appropriate server. */
+		}
+		if (tfw_http_adjust_req(req))
+			goto send_404;		/* Should it be an HTTP 502? */
+
+		/* Add request to the connection. */
+		list_add_tail(&req->msg.msg_list, &conn->msg_queue);
+
+		/* Send request to the server. */
```

Comment:
A request is processed in some NUMA node while the corresponding cached response can live in different NUMA node. So to minimize memory transfers we delegate request processing to appropriate NUMA node by deferred work in tfw_cache_req_process(). This asynchronous processing has nothing with _new_ response processing if cache can't fulfill current request. If we don't find a cached entity, then we should decide which NUMA node should create and deliver a response and schedule a work for that node. Meantime, we should do as less as possible work with the request fields on the remove node.

So moving the logic from tfw_http_req_process() to tfw_http_req_cache_cb() is incorrect.